app/controllers/listing_controller.rb

```diff
@@ -17,11 +17,38 @@ def use_react_app
     true
   end
 
-  def check_fcfs_flag
-    return if Rails.configuration.unleash.is_enabled? 'FCFS'
+  def should_allow_render
+    # if the FCFS feature flag is enabled
+    listing_id = params['id']
+    listing = Force::ListingService.listing(listing_id)
+    return if fcfs_flag_enabled &&
+              listing_type_fcfs_sales_bmr(listing) &&
+              listing_active(listing)
 
     # use the listing_path helper to redirect to the listing page
     # https://edgeguides.rubyonrails.org/routing.html#generating-paths-and-urls-from-code
     redirect_to listing_path(lang: params[:lang]) and return
   end
+
+  private
+
+  def fcfs_flag_enabled
+    Rails.configuration.unleash.is_enabled? 'FCFS'
+  end
+
+  def listing_type_fcfs_sales_bmr(listing)
+    # Record Type has to be Ownership
+    # Listing Type has to be First Come, First Served
+    listing['RecordType']['Name'] == 'Ownership' &&
+      listing['Listing_Type'] == 'First Come, First Served'
+  end
+
+  def listing_active(listing)
+    # Status has to be Active
+    # TODO: DAH-2846 Status will be added to the listing object
+    # Until then, that field will be nil
+    (!listing['Status'].nil? && listing['Status'] == 'Active') ||
+      # Accepting Online Applications has to be true
+      listing['Accepting_Online_Applications'] == true
+  end
```

Comment:
suggestion: As we are adding more logic to this controller I think we should either add tests for this controller or pull these out to files that we can better test

Reply:
To add more support this suggestion, a common pattern in Rails is to have ["skinny" controllers](https://thoughtbot.com/blog/skinny-controllers-skinny-models), i.e. almost zero business logic.

Reply:
The Rails [constraints](https://guides.rubyonrails.org/routing.html#advanced-constraints) feature may be helpful here.

Reply:
On it!